src/dune_rules/gen_rules.ml

```diff
@@ -438,10 +420,36 @@ let gen_rules ~sctx ~dir components : Build_config.gen_rules_result Memo.t =
         match
           String.Map.find automatic_sub_dirs_map (Path.Source.basename src_dir)
         with
-        | None -> redirect_to_parent
         | Some kind ->
           has_rules Subdir_set.empty (fun () ->
-              gen_rules_for_automatic_sub_dir ~sctx ~dir kind)))
+              gen_rules_for_automatic_sub_dir ~sctx ~dir kind)
+        | None -> (
+          let parent = Path.Build.parent_exn dir in
+          let+ stanzas = Only_packages.stanzas_in_dir parent in
+          match
+            match stanzas with
+            | None -> None
+            | Some stanzas ->
+              List.find_map stanzas.stanzas ~f:(function
+                | Melange_emit mel ->
+                  let target_dir = Path.Build.relative parent mel.target in
+                  Option.some_if (Path.Build.equal target_dir dir) mel
```

Comment:
I understand the purpose of the new code is to generate js rules from the directory where the melange.emit stanza lives (right?). But what is the code in these two lines doing specifically? Why do we compare `target_dir` with `dir`?

Reply:
> I understand the purpose of the new code is to generate js rules from the directory where the melange.emit stanza lives (right?). 

Actually, it is to generate the js rules from target directory of the melange stanza. So if we have `(melange.emit (target foo))` in `src/`, we'd like to generate the rules in `src/foo`.

> But what is the code in these two lines doing specifically? Why do we compare target_dir with dir?

This is where we detect the case above.